cpp/src/arrow/compute/exec/aggregate_node.cc

```diff
@@ -175,18 +175,21 @@ class ScalarAggregateNode : public ExecNode {
     return Status::OK();
   }
 
-  void InputReceived(ExecNode* input, ExecBatch batch) override {
+  void InputReceived(ExecNode* input, std::function<Result<ExecBatch>()> task) override {
     DCHECK_EQ(input, inputs_[0]);
 
     auto thread_index = get_thread_index_();
-
-    if (ErrorIfNotOk(DoConsume(std::move(batch), thread_index))) return;
+    auto prev = task();
+    if (!prev.ok()) {
+      ErrorIfNotOk(prev.status());
+      return;
+    }
+    if (ErrorIfNotOk(DoConsume(prev.MoveValueUnsafe(), thread_index))) return;
```

Comment:
```suggestion
    auto func = [this] (Result<ExecBatch> task) {
      ARROW_ASSIGN_OR_RAISE(auto prev, task());
      auto thread_index = get_thread_index_();
      return DoConsume(prev.MoveValueUnsafe(), thread_index);
    };
    plan_->scheduler()->SubmitTask(std::move(func));
```

This is what I'm thinking pipeline breakers would look like.

Reply:
> plan_->scheduler()->SubmitTask(std::move(func));

Yes that is the idea, but this PR is to enable that construction later, this PR is not going to define any scheduler or submitting logic.